protos/telemetry/telemetry.proto

```diff
@@ -564,11 +564,37 @@ enum FixType {
     FIX_TYPE_RTK_FIXED = 6; // RTK Fixed, 3D position
 }
 
+
+enum BatteryStatusFlags {
+    BATTERY_STATUS_FLAGS_READY_TO_USE = 1; // The battery is ready to use (fly).
+    BATTERY_STATUS_FLAGS_CHARGING = 2; // Battery is charging.
+    BATTERY_STATUS_FLAGS_CELL_BALANCING = 4; // Battery is cell balancing (during charging).
+    BATTERY_STATUS_FLAGS_FAULT_CELL_IMBALANCE = 8; // Battery cells are not balanced.
+    BATTERY_STATUS_FLAGS_AUTO_DISCHARGING = 16; // Battery is auto discharging (towards storage level).
+    BATTERY_STATUS_FLAGS_REQUIRES_SERVICE = 32; // Battery requires service (not safe to fly). 
+    BATTERY_STATUS_FLAGS_BAD_BATTERY = 64; // Battery is faulty and cannot be repaired (not safe to fly).
+    BATTERY_STATUS_FLAGS_PROTECTIONS_ENABLED = 128; // Automatic battery protection monitoring is enabled. 
+    BATTERY_STATUS_FLAGS_FAULT_PROTECTION_SYSTEM = 256; // The battery fault protection system had detected a fault and cut all power from the battery.
+    BATTERY_STATUS_FLAGS_FAULT_OVER_VOLT = 512; // One or more cells are above their maximum voltage rating.
+    BATTERY_STATUS_FLAGS_FAULT_UNDER_VOLT = 1024; // One or more cells are below their minimum voltage rating.
+    BATTERY_STATUS_FLAGS_FAULT_OVER_TEMPERATURE = 2048; // Over-temperature fault. 
+    BATTERY_STATUS_FLAGS_FAULT_UNDER_TEMPERATURE = 4096; // Under-temperature fault.
+    BATTERY_STATUS_FLAGS_FAULT_OVER_CURRENT = 8192; // Over-current fault.
+    BATTERY_STATUS_FLAGS_FAULT_SHORT_CIRCUIT = 16384; // Short circuit event detected.
+    BATTERY_STATUS_FLAGS_FAULT_INCOMPATIBLE_VOLTAGE = 32768; // Voltage not compatible with power rail voltage (batteries on same power rail should have similar voltage).
+    BATTERY_STATUS_FLAGS_FAULT_INCOMPATIBLE_FIRMWARE = 65536; // Battery firmware is not compatible with current autopilot firmware.
+    BATTERY_STATUS_FLAGS_FAULT_INCOMPATIBLE_CELLS_CONFIGURATION = 131072; // Battery is not compatible due to cell configuration.
+    BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL = 262144; // Battery capacity values are relative to a known-full battery (smart battery) and not estimated.
+    BATTERY_STATUS_FLAGS_EXTENDED = 4294967295; // Reserved for future use.
```

Comment:
This ID number is too high for protobuf (hence the prototool failure in the CI) :grin:

Reply:
We generally don't support enums used as bitflags in MAVSDK, right @JonasVautherin?

So what we can do is add boolean values for it and then convert between the flags and bools.
I think this makes using the API easier for other languages, where it's not as common to do bitwise checks.

Reply:
Well as long as the number is in bounds, it's fine. Usually we just increment from 0, but well... I don't really mind if we want to use weird ids :sweat_smile:.

However the protobuf field number is only a protobuf thing, it's completely independent from MAVLink. So using powers of two has no meaning here :+1:

Reply:
So that's the thing, we don't support it because the numbers are removed in the C++ API.

Reply:
Sorry I did not get that. In the C++ implementation, the code will have to translate this enum into the mavlink message. But using powers of 2 here doesn't do anything, so it would probably be cleaner to just increment from 0.

Reply:
I thought I was defining an enum with preset values, but it sounds like you're saying that the value I am assigning in (say) ` BATTERY_STATUS_FLAGS_FAULT_OVER_CURRENT = 8192;` is a protobuf field number.

So what the hell is the resulting enum value, and how would I use it in a bitmask? My best-guess would be that (if the number is just an internal protobuf thing) the enums would copy across to whatever language in whatever form it understands as an auto-numbered enum from zero. 
The implication being that you CANT use these as flags right?

So what I "think" you (@julianoes) are saying is that I can't expose enums as bitmaps through the API. I need to create bools for each flag. Right?
So in the proto I'd get rid of these enums and it might look like:

```
bool status_not_ready_to_use= 5  [(mavsdk.options.default_value)="0"]; // Battery status: not ready to use
bool status_charging= 6  [(mavsdk.options.default_value)="0"]; // Battery status: charging
...
````

I'm right when I assume that these definitions aren't described in the docs right? (checked via search, perhaps I just missed it)?

Reply:
> So what the hell is the resulting enum value, and how would I use it in a bitmask?

That's an implementation detail, that's not the responsibility of the API. Using bitmasks is an optimization at the protocol level, it has nothing to do here.

The part of the code that will create and send the actual mavlink message will understand the enum, and convert it to whatever mavlink wants :slightly_smiling_face:.

> I need to create bools for each flag. Right?

No, the enum is fine. As a MAVSDK user, I'd rather deal with `BATTERY_STATUS_FLAGS_EXTENDED` than `4294967295`. I don't care how the enum is encoded on the wire, be it with a bitmask, with booleans, little-endian or not, UTF8 or not, etc.

> I'm right when I assume that these definitions aren't described in the docs right?

The proto definition is documented e.g. [here](https://developers.google.com/protocol-buffers/docs/proto3) if you are interested. But let's be honest, nobody ever reads that. Instead they look at the other MAVSDK protos, write their version, and let us maintainers tell them if that's okay or not.

Also I believe that if you tried to generate the code and implement the C++ code, you would see that this API gives you an enum and that you need to translate it where you create the mavlink message :innocent:.

Reply:
> No, the enum is fine. As a MAVSDK user, I'd rather deal with BATTERY_STATUS_FLAGS_EXTENDED than 4294967295. I don't care how the enum is encoded on the wire, be it with a bitmask, with booleans, little-endian or not, UTF8 or not, etc.

@JonasVautherin I don't think this would work. The problem is that the MAVLink spec is not really an enum but it's bitflags. 

So in our conversion the enum becomes a normal enum incremented with 1 (instead of doubled for bits), so you can only ever set one flag at a time. In the MAVLink world you can set multiple bits.

Therefore, I'm suggesting to use boolean flags for each flag, and then encode it to bits internally.

Reply:
Ooooooh, I get it now :+1:. Yeah then a boolean makes total sense :100:

Reply:
So is the format I suggested right?

```
bool status_not_ready_to_use= 5  [(mavsdk.options.default_value)="0"]; // Battery status: not ready to use
bool status_charging= 6  [(mavsdk.options.default_value)="0"]; // Battery status: charging
...
```

Reply:
I would try with `[(mavsdk.options.default_value)="false"]`, I think that should work :innocent:

Reply:
Yes, that looks right.